File: tools/end_call.go
package tools

import (
	"context"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
	"jordandavis.dev/phone-call-mcp-server/clients"
)

var endCallDef = &mcp.Tool{
	Name:        "EndCall",
	Description: "Ends an active phone call by its call ID.",
}

type endCallInput struct {
	CallId string `json:"callId" jsonschema:"The ID of the call to end."`
}

type endCallOutput struct {
	Result string
	Err    error
}

func EndCall(ctx context.Context, req *mcp.CallToolRequest, input endCallInput) (
	*mcp.CallToolResult,
	endCallOutput,
	error,
) {
	callId := input.CallId

	client := clients.NewTwilioClient()

	err := client.EndCall(callId)
	if err != nil {
		return nil, endCallOutput{}, err
	}

	result := fmt.Sprintf("Call %s successfully ended.", callId)

	return nil, endCallOutput{result, nil}, nil
}

func AddToolEndCall(server *mcp.Server) {
	mcp.AddTool(server, endCallDef, EndCall)
}

File: src/tools/endCall.ts
import type { McpServer, ToolCallback } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import type { EndCallClient } from "../clients/twilioClient";

const inputSchema = {
  callId: z.string().min(1).describe("The ID of the call to end."),
};

export function registerEndCallTool(server: McpServer, client: EndCallClient): void {
  const handler: ToolCallback<typeof inputSchema> = async ({ callId }) => {
    try {
      await client.endCall(callId);
      const result = { success: true };
      return {
        content: [{ type: "text", text: "Call successfully ended." }],
        structuredContent: result,
      };
    } catch {
      return {
        content: [{ type: "text", text: "Failed to end call." }],
        isError: true,
      };
    }
  };

  server.registerTool(
    "EndCall",
    {
      description: "Ends an active phone call by its call ID.",
      inputSchema,
    },
    handler,
  );
}
